We import necessary libaries first

In [121]:
import yfinance as yf
yf.pdr_override()
import pandas as pd
import numpy as np
from scipy.stats import kurtosis, skew
from sklearn.model_selection import train_test_split

### a) Download 1-2 years of price history of a stock.



In [75]:
Apple= yf.download("AAPL", start="2019-09-01", end="2020-09-01")

[*********************100%***********************]  1 of 1 completed


In [76]:
Apple.head()

Open       High        Low      Close  Adj Close     Volume
Date                                                                        
2019-09-03  51.607498  51.744999  51.055000  51.424999  50.919777   80092000
2019-09-04  52.097500  52.369999  51.830002  52.297501  51.783714   76752400
2019-09-05  53.000000  53.492500  52.877499  53.320000  52.796162   95654800
2019-09-06  53.512501  53.605000  53.127499  53.314999  52.791210   77449200
2019-09-09  53.709999  54.110001  52.767502  53.542500  53.016479  109237600

### b) Compute its log return.

In [77]:
Apple["Log_Return_Apple"]=np.log(Apple["Close"]/Apple["Close"].shift(1))

We do not need every columns in this datasets, so i droped some of it

In [78]:
Apple.drop(['Open', 'High', 'Low','Close', 'Adj Close', 'Volume'], axis=1)

Log_Return_Apple
Date                        
2019-09-03               NaN
2019-09-04          0.016824
2019-09-05          0.019363
2019-09-06         -0.000094
2019-09-09          0.004258
...                      ...
2020-08-25         -0.008238
2020-08-26          0.013507
2020-08-27         -0.012026
2020-08-28         -0.001621
2020-08-31          0.033350

[252 rows x 1 columns]

### c) Compute the mean, standard deviation, skewness, and excess kurtosis of its log return. 

In [79]:
print( 'Mean of Log Return of Apple stocks: {}'.format(np.mean(Apple["Log_Return_Apple"].dropna())))
print( 'Standard Deviation of Log Return of Apple stocks: {}'.format(np.std(Apple["Log_Return_Apple"].dropna())))
print( 'Excess kurtosis of Log Return of Apple stocks(should be 0): {}'.format(kurtosis(Apple["Log_Return_Apple"].dropna())))
print( 'Skewness of Log Return of Apple stocks (should be 0): {}'.format( skew(Apple["Log_Return_Apple"].dropna())))

Mean of Log Return of Apple stocks: 0.0036653305222491927
Standard Deviation of Log Return of Apple stocks: 0.02631950821630945
Excess kurtosis of Log Return of Apple stocks(should be 0): 6.504365071806104
Skewness of Log Return of Apple stocks (should be 0): -0.34483332400211414


### d) Repeat for a second stock.  

In [80]:
Tesla= yf.download("TSLA", start="2019-09-01", end="2020-09-01")

[*********************100%***********************]  1 of 1 completed


In [81]:
Tesla.head()

Open       High        Low      Close  Adj Close    Volume
Date                                                                       
2019-09-03  44.816002  45.790001  44.632000  45.001999  45.001999  26770500
2019-09-04  45.377998  45.692001  43.841999  44.136002  44.136002  28805000
2019-09-05  44.500000  45.959999  44.169998  45.916000  45.916000  36976500
2019-09-06  45.439999  45.928001  45.034000  45.490002  45.490002  20947000
2019-09-09  46.000000  46.751999  45.846001  46.358002  46.358002  24013500

In [82]:
Tesla["Log_Return_Tesla"]=np.log(Tesla["Close"]/Tesla["Close"].shift(1))

In [83]:
Tesla.drop(['Open', 'High', 'Low', 'Close','Adj Close', 'Volume'], axis=1)

Log_Return_Tesla
Date                        
2019-09-03               NaN
2019-09-04         -0.019431
2019-09-05          0.039538
2019-09-06         -0.009321
2019-09-09          0.018901
...                      ...
2020-08-25          0.004528
2020-08-26          0.062192
2020-08-27          0.038976
2020-08-28         -0.011388
2020-08-31          0.118395

[252 rows x 1 columns]

In [84]:
print( 'Mean of Log Return of Tesla stocks: {}'.format(np.mean(Tesla["Log_Return_Tesla"].dropna())))
print( 'Standard Deviation of Log Return of Tesla stocks: {}'.format(np.std(Tesla["Log_Return_Tesla"].dropna())))
print( 'excess kurtosis of Log Return of Tesla stocks(should be 0): {}'.format(kurtosis(Tesla["Log_Return_Tesla"].dropna())))
print( 'skewness of Log Return of Tesla stocks (should be 0): {}'.format( skew(Tesla["Log_Return_Tesla"].dropna()) ))

Mean of Log Return of Tesla stocks: 0.009579822895735262
Standard Deviation of Log Return of Tesla stocks: 0.05135007432639206
excess kurtosis of Log Return of Tesla stocks(should be 0): 3.3461653005457572
skewness of Log Return of Tesla stocks (should be 0): -0.2312716159720003


### e) Compute the covariance and the correlation. Explain their difference. How do you convert one to the other?

In [85]:
print( 'Covariance of Log Return of Apple and Tesla Stocks: {}'.format( Apple["Log_Return_Apple"].cov(Tesla["Log_Return_Tesla"])))

Covariance of Log Return of Apple and Tesla Stocks: 0.0006266850826014723


In [86]:
print( 'Correlation between Log Return of Apple and Tesla Stocks: {}'.format( Apple["Log_Return_Apple"].corr(Tesla["Log_Return_Tesla"])))

Correlation between Log Return of Apple and Tesla Stocks: 0.4618455862203561


Both correlation and covariance measure the relationship and the dependency between two variables.Here, our two variables are Apple's  and Tesla's stocks. Covariance is used to find how much Apple and Tesla stocks vary together, whereas correlation is used to find when a change in one of our variable, say Apple can result in a change in our another variable which is Tesla.

As the formula for Correlation $$Corr=\frac{Cov\left ( Apple, Tesla \right )}{\sigma _{Apple}^{2}*\sigma _{Tesla}^{2}}$$

We will be able to convert between them by using abovementioned formula

# 2. Build your own transition

### a) Divide the data into 2 uneven parts: the first part is 80% of your data, and the second part is 20%. 

In [87]:
train_apple, test_apple = train_test_split(Apple, test_size=0.2,train_size=0.8, random_state=42, shuffle=True)

We can check the length of splitted data as below

In [92]:
print(len(train_apple))
print(len(test_apple))

201
51


80% of Data

In [93]:
train_apple.head()

Open       High        Low      Close  Adj Close     Volume  \
Date                                                                           
2020-03-18  59.942501  62.500000  59.279999  61.667500  61.390331  300233600   
2019-11-19  66.974998  67.000000  66.347504  66.572502  66.116371   76167200   
2020-06-24  91.250000  92.197502  89.629997  90.014999  89.852997  192623200   
2020-07-13  97.264999  99.955002  95.257500  95.477501  95.305664  191649200   
2020-04-02  60.084999  61.287498  59.224998  61.232498  60.957287  165934000   

            Log_Return_Apple  
Date                          
2020-03-18         -0.024785  
2019-11-19         -0.003037  
2020-06-24         -0.017810  
2020-07-13         -0.004624  
2020-04-02          0.016549

20% of Data

In [94]:
test_apple.head()

Open       High        Low      Close  Adj Close     Volume  \
Date                                                                           
2020-04-29  71.182503  72.417503  70.972504  71.932503  71.609200  137280800   
2019-09-11  54.517502  55.927502  54.432499  55.897499  55.348343  177158400   
2020-02-11  80.900002  80.974998  79.677498  79.902496  79.543365   94323200   
2020-05-08  76.410004  77.587502  76.072502  77.532501  77.392960  134048000   
2020-02-18  78.839996  79.937500  78.652496  79.750000  79.391556  152531200   

            Log_Return_Apple  
Date                          
2020-04-29          0.032317  
2019-09-11          0.031300  
2020-02-11         -0.006052  
2020-05-08          0.020820  
2020-02-18         -0.018480

Now, let's do the same for the second stock and divide it into 80% and 20% datasets

In [95]:
train_tesla, test_tesla = train_test_split(Tesla, test_size=0.2,train_size=0.8, random_state=42, shuffle=True)

We can check the length of splitted data as below

In [97]:
print(len(train_tesla))
print(len(test_tesla))

201
51


80% of Data

In [98]:
train_tesla.head()

Open        High         Low       Close   Adj Close  \
Date                                                                     
2020-03-18   77.800003   80.972000   70.101997   72.244003   72.244003   
2019-11-19   70.349998   71.998001   69.559998   71.903999   71.903999   
2020-06-24  198.822006  200.175995  190.628006  192.169998  192.169998   
2020-07-13  331.799988  358.997986  294.221985  299.411987  299.411987   
2020-04-02   96.206001   98.851997   89.279999   90.893997   90.893997   

               Volume  Log_Return_Tesla  
Date                                     
2020-03-18  118931000         -0.174763  
2019-11-19   38624000          0.026865  
2020-06-24   54798000         -0.041715  
2020-07-13  194927000         -0.031294  
2020-04-02   99292000         -0.057899

20% of Data

In [99]:
test_tesla.head()

Open        High         Low       Close   Adj Close  \
Date                                                                     
2020-04-29  158.033997  160.639999  156.632004  160.102005  160.102005   
2019-09-11   47.476002   49.633999   47.200001   49.419998   49.419998   
2020-02-11  153.757996  156.701996  151.600006  154.876007  154.876007   
2020-05-08  158.753998  164.800003  157.401993  163.884003  163.884003   
2020-02-18  168.320007  172.000000  166.472000  171.679993  171.679993   

              Volume  Log_Return_Tesla  
Date                                    
2020-04-29  81080000          0.040002  
2019-09-11  50214000          0.047912  
2020-02-11  58487500          0.004011  
2020-05-08  80650500          0.049252  
2020-02-18  81908500          0.070421

### b) Categorize each day in the 1-2 year price history as belonging to one of four categories:
i.Both stocks up
ii.Stock 1 up, stock 2 down
iii.Stock 1 down, stock 2 up
iv.Both stocks down

Let's drop unnecessary columns in two datasets first and conctenate what remains in both as one single datasets

In [100]:
df1=Apple.drop(['Open', 'High', 'Low','Close', 'Adj Close', 'Volume'], axis=1)
df2=Tesla.drop(['Open', 'High', 'Low', 'Close','Adj Close', 'Volume'], axis=1)
frames=[df1,df2]
Log_Returns_Cat=pd.concat(frames,  axis=1)

In [101]:
Log_Returns_Cat

Log_Return_Apple  Log_Return_Tesla
Date                                          
2019-09-03               NaN               NaN
2019-09-04          0.016824         -0.019431
2019-09-05          0.019363          0.039538
2019-09-06         -0.000094         -0.009321
2019-09-09          0.004258          0.018901
...                      ...               ...
2020-08-25         -0.008238          0.004528
2020-08-26          0.013507          0.062192
2020-08-27         -0.012026          0.038976
2020-08-28         -0.001621         -0.011388
2020-08-31          0.033350          0.118395

[252 rows x 2 columns]

To categorize as it is mentioned above, we need to get the difference between consecutive stocks for each data and see if it goes up or down.If the difference between todays and tomorrows stock is negative  then there is a down otherwise it is up(there is no 0 case)

In [103]:
Diff_Return=Log_Returns_Cat.diff()
Diff_Return


Log_Return_Apple  Log_Return_Tesla
Date                                          
2019-09-03               NaN               NaN
2019-09-04               NaN               NaN
2019-09-05          0.002539          0.058969
2019-09-06         -0.019457         -0.048859
2019-09-09          0.004352          0.028222
...                      ...               ...
2020-08-25         -0.020127          0.022135
2020-08-26          0.021745          0.057664
2020-08-27         -0.025534         -0.023215
2020-08-28          0.010405         -0.050364
2020-08-31          0.034971          0.129783

[252 rows x 2 columns]

Now, we can label each stock as up and and down depending on difference between consecutive days.

In [106]:
# create a list of our conditions
conditions = [
    (Diff_Return['Log_Return_Apple']>0)& (Diff_Return['Log_Return_Tesla']>0),
    (Diff_Return['Log_Return_Apple']>0) & (Diff_Return['Log_Return_Tesla']<0),
    (Diff_Return['Log_Return_Apple']<0) & (Diff_Return['Log_Return_Tesla']>0),
    (Diff_Return['Log_Return_Apple']<0) & (Diff_Return['Log_Return_Tesla']<0)
    ]

# create a list of the values we want to assign for each condition
values = ['Both stocks up', 'Stock #1 up, stock #2 down', 'Stock #1 down, stock #2 up', 'Both stocks down']

# create a new column and use np.select to assign values to it using our lists as arguments
Log_Returns_Cat['Categories'] = np.select(conditions, values)

# display updated DataFrame
Log_Returns_Cat


Log_Return_Apple  Log_Return_Tesla                  Categories
Date                                                                      
2019-09-03               NaN               NaN                           0
2019-09-04          0.016824         -0.019431                           0
2019-09-05          0.019363          0.039538              Both stocks up
2019-09-06         -0.000094         -0.009321            Both stocks down
2019-09-09          0.004258          0.018901              Both stocks up
...                      ...               ...                         ...
2020-08-25         -0.008238          0.004528  Stock #1 down, stock #2 up
2020-08-26          0.013507          0.062192              Both stocks up
2020-08-27         -0.012026          0.038976            Both stocks down
2020-08-28         -0.001621         -0.011388  Stock #1 up, stock #2 down
2020-08-31          0.033350          0.118395              Both stocks up

[252 rows x 3 columns]

### c. Build a transition matrix of portfolio direction that shows your portfolio in          four scenarios:


i.  From moving together to moving together That means starting from uu or dd & going to uu or dd
ii. From moving together to moving apart That means starting from uu or dd & going to ud or du
iii. From moving apart to moving together That means starting from ud or du & going to uu or dd
iv. From moving apart to moving apart That means starting from ud or du & going to ud or du  

In [107]:
Log_Returns_Cat.drop(Log_Returns_Cat.index[0])
Log_Returns_Cat.groupby(by=['Categories']).count()

Log_Return_Apple  Log_Return_Tesla
Categories                                                    
0                                          1                 1
Both stocks down                          79                79
Both stocks up                            77                77
Stock #1 down, stock #2 up                47                47
Stock #1 up, stock #2 down                47                47

Let's create a sequence of all uu, dd,ud and dd

In [113]:
given_value1 ='dd'
new_list1=[]
new_list1.extend([given_value1 for i in range(79)])
given_value2 ='uu'
new_list2=[]
new_list2.extend([given_value2 for i in range(77)])
given_value3 ='du'
new_list3=[]
new_list3.extend([given_value3 for i in range(47)])
given_value4 ='ud'
new_list4=[]
new_list4.extend([given_value4 for i in range(47)])

In [114]:
sequence=new_list1 + new_list2 + new_list3 + new_list4

In [110]:
from itertools import islice

def window(seq, n=2):
    it = iter(seq)
    result = tuple(islice(it, n))
    if len(result) == n:
        yield result
    for elem in it:
        result = result[1:] + (elem,)
        yield result

Let's create new dataframe and name the columns as state 1 and state 2 and by usuing our window() function( i created above)
Let's put our pairs in new dataframe, afterthat Let's get the probability of each state 1 changing to state 2

In [115]:
pairs = pd.DataFrame(window(sequence), columns=['state1', 'state2'])
counts = pairs.groupby('state1')['state2'].value_counts()
probs = (counts / counts.sum()).unstack()

Let's get our probabality in a new dataframe all together

In [117]:
df=pd.DataFrame(probs)
df

state2        dd        du        ud        uu
state1                                        
dd      0.313253       NaN       NaN  0.004016
du           NaN  0.184739  0.004016       NaN
ud           NaN       NaN  0.184739       NaN
uu           NaN  0.004016       NaN  0.305221

Let's slice our matrix as question asks 

In [119]:
df.loc[['dd','uu'],['dd','uu']]

state2        dd        uu
state1                    
dd      0.313253  0.004016
uu           NaN  0.305221

In [120]:
df.loc[['du','ud'],['du','ud']]

state2        du        ud
state1                    
du      0.184739  0.004016
ud           NaN  0.184739

### d. Discuss the similarities or differences of the two transition matrices  

Both matrix have one NaN value, which is uu to dd for first and ud to du for second matrix. Probability of dd to  dd is the greatest probability among them with 0.313 and uu to uu follows it as  second biggest with 0.310. It looks like dd to uu and du to ud has same probability of happening. 

### e. Is the process Markovian?

Yes, this is Markovian as it is memoryless and we can predict future outcome based solely on its present state, we do not need its full history.